File: frontend/types/scan.ts
import type { PredictionItem } from "@/types/prediction";

export type ScanRecord = {
  id: string;
  session_id: string;
  image_url?: string | null;
  gradcam_url?: string | null;
  lime_url?: string | null;
  prediction?: string | null;
  raw_label?: string | null;
  confidence?: number | null;
  is_confident?: boolean | null;
  weather_risk?: string | null;
  weather_risk_score?: number | null;
  combined_risk_score?: number | null;
  temperature?: number | null;
  humidity?: number | null;
  rainfall?: number | null;
  latitude?: number | null;
  longitude?: number | null;
  location_name?: string | null;
  top_predictions?: PredictionItem[];
  explanation?: string | null;
  next_steps?: string[];
  disclaimer?: string | null;
  created_at?: string | null;
};

export type SaveScanRequest = {
  session_id: string;
  original_image?: string | null;
  gradcam_image?: string | null;
  lime_image?: string | null;
  prediction: string;
  raw_label?: string | null;
  confidence?: number | null;
  is_confident?: boolean | null;
  weather_risk?: string | null;
  weather_risk_score?: number | null;
  combined_risk_score?: number | null;
  temperature?: number | null;
  humidity?: number | null;
  rainfall?: number | null;
  latitude?: number | null;
  longitude?: number | null;
  location_name?: string | null;
  top_predictions?: PredictionItem[];
  explanation?: string | null;
  next_steps?: string[];
  disclaimer?: string | null;
};

export type SaveScanResponse = {
  scan: ScanRecord;
};

File: backend/app/schemas/scan.py
from typing import Any

from pydantic import BaseModel, Field


class TopPrediction(BaseModel):
    label: str
    raw_label: str
    confidence: float


class ScanCreate(BaseModel):
    session_id: str = Field(min_length=1)
    original_image: str | None = None
    gradcam_image: str | None = None
    lime_image: str | None = None
    prediction: str
    raw_label: str | None = None
    confidence: float | None = None
    is_confident: bool | None = None
    weather_risk: str | None = None
    weather_risk_score: float | None = None
    combined_risk_score: float | None = None
    temperature: float | None = None
    humidity: float | None = None
    rainfall: float | None = None
    latitude: float | None = None
    longitude: float | None = None
    location_name: str | None = None
    top_predictions: list[TopPrediction] = Field(default_factory=list)
    explanation: str | None = None
    next_steps: list[str] = Field(default_factory=list)
    disclaimer: str | None = None


class ScanRecord(BaseModel):
    id: str
    session_id: str
    image_url: str | None = None
    gradcam_url: str | None = None
    lime_url: str | None = None
    prediction: str | None = None
    raw_label: str | None = None
    confidence: float | None = None
    is_confident: bool | None = None
    weather_risk: str | None = None
    weather_risk_score: float | None = None
    combined_risk_score: float | None = None
    temperature: float | None = None
    humidity: float | None = None
    rainfall: float | None = None
    latitude: float | None = None
    longitude: float | None = None
    location_name: str | None = None
    top_predictions: list[dict[str, Any]] = Field(default_factory=list)
    explanation: str | None = None
    next_steps: list[str] = Field(default_factory=list)
    disclaimer: str | None = None
    created_at: str | None = None


class ScanSaveResponse(BaseModel):
    scan: ScanRecord
